modules/renter/repairscanner.go

```diff
@@ -0,0 +1,333 @@
+	memoryNeeded     uint64 // memory needed in bytes
+	piecesNeeded     int // number of pieces to achieve a 100% complete upload
+	piecesCompleted  int // number of pieces that have been fully uploaded.
+	piecesRegistered int // number of pieces that are being uploaded, but aren't finished yet.
+	pieceUsage       []bool // one per piece. 'false' = piece not uploaded. 'true' = piece uploaded.
+	progress         float64 // percent complete at the moment uploading started
+	unusedHosts      map[string]struct{} // hosts that aren't yet storing any pieces
+
+	// Utilities.
+	mu sync.Mutex
+}
+
+func (ch chunkHeap) Len() int            { return len(ch) }
+func (ch chunkHeap) Less(i, j int) bool  { return ch[i].progress < ch[j].progress }
+func (ch chunkHeap) Swap(i, j int)       { ch[i], ch[j] = ch[j], ch[i] }
+func (ch *chunkHeap) Push(x interface{}) { *ch = append(*ch, x.(*unfinishedChunk)) }
+func (ch *chunkHeap) Pop() interface{} {
+	old := *ch
+	n := len(old)
+	x := old[n-1]
+	*ch = old[0 : n-1]
+	return x
+}
+
+// buildUnfinishedChunks will pull all of the unfinished chunks out of a file.
+//
+// TODO / NOTE: This code can be substantially simplified once the files store
+// the HostPubKey instead of the FileContractID, and can be simplified even
+// further once the layout is per-chunk instead of per-filecontract.
+func (r *Renter) buildUnfinishedChunks(f *file, hosts map[string]struct{}, fcidToHPK map[types.FileContractID]types.SiaPublicKey) []*unfinishedChunk {
+	// Files are not threadsafe.
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	// If the file is not being tracked, don't repair it.
+	trackedFile, exists := r.tracking[f.name]
+	if !exists {
+		return nil
+	}
+
+	// Assemble the set of chunks.
+	//
+	// TODO / NOTE: Future files may have a different method for determining the
+	// number of chunks. Changes will be made due to things like sparse files,
+	// and the fact that chunks are going to be different sizes.
+	chunkCount := f.numChunks()
+	newUnfinishedChunks := make([]*unfinishedChunk, chunkCount)
+	// Add a separate unusedHosts map for each chunk, as every chunk will have a
+	// different set of unused hosts.
+	for i := uint64(0); i < chunkCount; i++ {
+		newUnfinishedChunks[i] = new(unfinishedChunk)
+		newUnfinishedChunks[i].index = i
+		newUnfinishedChunks[i].localPath = trackedFile.RepairPath
+
+		// Mark the number of pieces needed for this chunk.
+		newUnfinishedChunks[i].piecesNeeded = f.erasureCode.NumPieces()
+		newUnfinishedChunks[i].memoryNeeded = f.pieceSize * uint64(f.erasureCode.NumPieces())
+
+		// TODO / NOTE: Offset and length are going to have to be derived using
+		// alternate means once chunks are no longer constant size within a
+		// file. Likely the chunk metadata will contain this information, but we
+		// also want to make sure that files are random-access, and don't
+		// require seeking through a ton of chunk headers to get to an arbitrary
+		// position. It's currently an open problem.
```

Comment:
if the format makes it easy to load the `length`s of each chunk into memory, we can quickly seek by summing them and stopping once the sum exceeds the desired offset:
```go
func chunkAtOffset(chunkLens []uint64, offset uint64) (chunkIndex, offsetWithinChunk int) {
	rem := offset
	for i, s := range chunkLens {
		if rem < s {
			return i, int(rem)
		}
		rem -= s
	}
	return -1, -1
}
```

Reply:
if you have 10,000 chunks in a file though (admittedly, a big file), you don't want to have to do 10,000 iterations just to get the offset.

out of scope for this pr though

Reply:
Even 10k iterations is negligible -- takes about 4000 ns. The bigger issue would be storing the lengths. If they're loaded from disk each time, that adds non-trivial overhead; if they're stored in memory, then they add to the RAM weight of the renter. Either way, we'll have bigger fish to fry wrt scaling issues.